src/main/kotlin/uk/gov/justice/digital/hmpps/incidentreporting/dto/nomis/Mapping.kt

```diff
@@ -43,93 +46,71 @@ fun NomisReport.toNewEntity(): Report {
-  }
-}
+private fun Report.createStaffInvolved(staffParty: NomisStaffParty): StaffInvolvement =
+  StaffInvolvement(
+    report = this,
+    staffUsername = staffParty.staff.username,
+    staffRole = StaffRole.fromNomisCode(staffParty.role.code),
+    comment = staffParty.comment,
+  )
 
-fun Report.addNomisPrisonerInvolvements(offenderParties: Collection<NomisOffenderParty>) {
-  offenderParties.forEach {
-    this.addPrisonerInvolved(
-      prisonerNumber = it.offender.offenderNo,
-      prisonerRole = PrisonerRole.fromNomisCode(it.role.code),
-      outcome = it.outcome?.let { prisonerOutcome -> PrisonerOutcome.fromNomisCode(prisonerOutcome.code) },
-      comment = it.comment,
-    )
-  }
-}
+private fun Report.createPrisonerInvolvement(nomisOffenderParty: NomisOffenderParty): PrisonerInvolvement =
+  PrisonerInvolvement(
+    report = this,
+    prisonerNumber = nomisOffenderParty.offender.offenderNo,
+    prisonerRole = PrisonerRole.fromNomisCode(nomisOffenderParty.role.code),
+    outcome = nomisOffenderParty.outcome?.let { prisonerOutcome -> PrisonerOutcome.fromNomisCode(prisonerOutcome.code) },
+    comment = nomisOffenderParty.comment,
+  )
 
-fun Report.addNomisCorrectionRequests(correctionRequests: Collection<NomisRequirement>) {
-  correctionRequests.forEach {
-    this.addCorrectionRequest(
-      correctionRequestedBy = it.staff.username,
-      correctionRequestedAt = it.date.atStartOfDay(),
-      descriptionOfChange = it.comment ?: NO_DETAILS_GIVEN,
-      reason = CorrectionReason.NOT_SPECIFIED,
-    )
-  }
-}
+private fun Report.createCorrectionRequest(correctionRequest: NomisRequirement): CorrectionRequest =
+  CorrectionRequest(
+    report = this,
+    correctionRequestedBy = correctionRequest.staff.username,
+    correctionRequestedAt = correctionRequest.date.atStartOfDay(),
+    descriptionOfChange = correctionRequest.comment ?: NO_DETAILS_GIVEN,
+    reason = CorrectionReason.NOT_SPECIFIED,
+  )
 
-fun Report.addNomisQuestions(questions: Collection<NomisQuestion>) {
-  questions.forEach { question ->
-    val dataItem = this.addQuestion(
-      code = question.questionId.toString(),
-      sequence = question.sequence - 1,
-      question = question.question,
-    )
-    question.answers
-      .forEach { answer ->
-        dataItem.addResponse(
-          response = answer.answer!!,
-          sequence = answer.sequence - 1,
-          responseDate = answer.responseDate,
-          additionalInformation = answer.comment,
-          recordedBy = answer.recordingStaff.username,
-          recordedAt = this.reportedAt,
-        )
-      }
-  }
+fun Report.updateNomisPrisonerInvolvements(offenderParties: Collection<NomisOffenderParty>) {
+  val newInvolvements = offenderParties.map { offenderParty ->
+    val newPrisoner = createPrisonerInvolvement(offenderParty)
+    prisonersInvolved.find { it == newPrisoner } ?: addPrisonerInvolved(newPrisoner)
+  }.toSet()
+
+  this.prisonersInvolved.retainAll(newInvolvements)
```

Comment:
are the retain-all calls a bit of a jpa fudge? the javadocs suggest that only elements *already* in the mutable set are kept. does it mean it's bit of an accident that any newly-created entity actually get saved?

or is the jpa/hibernate design such that instantiating an entity is enough for it to be inserted into db without needing to actually be in this sorted set?

Reply:
no it needs to be in the sorted set.  None of this is JPA, it just set theory

Reply:
I struggled to understand this at first but then I realised the `addPrisonerInvolved(newPrisoner)` bit was adding to `this.prisonersInvolved` rather than creating a DB record (it makes sense, but I didn't quite get that at first)

Reply:
right, but there was a point when the entities were [not being added](https://github.com/ministryofjustice/hmpps-incident-reporting-api/pull/226/commits/e6467fe4709c978d881dc32d4117eac8403761dc)

my question was _only_ about the use of `SortedSet.retainAll`